src/main/kotlin/com/openlattice/data/storage/PostgresQueries.kt

```diff
@@ -375,11 +378,23 @@ internal fun selectEntityKeyIdsWithCurrentVersionSubquerySql(
 ): String {
     val metadataColumns = getMetadataOptions(metadataOptions, linking).joinToString(",")
 
-    val selectColumns = joinColumns.joinToString(",") +
-            (if (!metadataColumns.isEmpty()) ", $metadataColumns" else "")
-    val distinct = if (linking) "DISTINCT" else "" // we need to add this in case of linking query, because linking_id, entity_set_id are not unique
 
-    return "(SELECT $distinct $selectColumns FROM ${IDS.name} WHERE ${VERSION.name} > 0 $entitiesClause ) as $ENTITIES_TABLE_ALIAS"
+    var selectColumns = joinColumns.joinToString(",")
+
+    if (metadataColumns.isNotEmpty()) {
+
+        if (linking) {
+            if (metadataOptions.contains(MetadataOption.LAST_WRITE)) {
+                selectColumns += ", max(${LAST_WRITE.name}) AS ${LAST_WRITE.name}"
+            }
+        } else {
+            selectColumns += ", $metadataColumns"
+        }
+    }
+
+    val groupBy = if (linking) "GROUP BY ${LINKING_ID.name}" else ""
+
+    return "(SELECT $selectColumns FROM ${IDS.name} WHERE ${VERSION.name} > 0 $entitiesClause $groupBy ) as $ENTITIES_TABLE_ALIAS"
```

Comment:
distinct is missing from the select, if query is for linking entities

Reply:
it's not needed anymore because of the groupBy

Reply:
I see, nvm then!